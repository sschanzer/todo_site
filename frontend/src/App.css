.checkHolder {
  display: flex;
  justify-content: center;
  align-items: top;
  position: relative;
}
.check {
  position: absolute;
  opacity: 0;
  z-index: 2;
}
.alterBtn {
  height: 3vh;
  width: 2vw;
  font-size: 2vh;
  text-align: center;
  display: flex !important;
  justify-content: center;
  align-items: center;
}

.buttonHolder {
  display: flex;
}

.buttonImg {
  height: 2vh;
}
