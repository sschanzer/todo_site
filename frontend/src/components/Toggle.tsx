import ToggleButton from "react-bootstrap/ToggleButton";
import ToggleButtonGroup from "react-bootstrap/ToggleButtonGroup";

interface ToggleProps {
  showPending: boolean;
  setShowPending: (showPending: boolean) => void;
  showCompleted: boolean;
  setShowCompleted: (showCompleted: boolean) => void;
}

export const Toggle: React.FC<ToggleProps> = ({
  showPending,
  setShowPending,
  showCompleted,
  setShowCompleted,
}) => {
  return (
    <ToggleButtonGroup defaultValue={[1, 3]} size="sm" type="checkbox">
      <ToggleButton
        variant="outline-primary"
        id="tbg-btn-1"
        checked={showPending}
        defaultChecked={showPending}
        onChange={(e) =>
          showCompleted
            ? setShowPending(e.target.checked)
            : [
                setShowPending(e.target.checked),
                setShowCompleted(!e.target.checked),
              ]
        }
        value={1}
      >
        Pending
      </ToggleButton>
      <ToggleButton
        variant="outline-primary"
        id="tbg-btn-3"
        checked={showCompleted}
        defaultChecked={showCompleted}
        onChange={(e) =>
          showPending
            ? setShowCompleted(e.target.checked)
            : [
                setShowCompleted(e.target.checked),
                setShowPending(!e.target.checked),
              ]
        }
        value={3}
      >
        Complete
      </ToggleButton>
    </ToggleButtonGroup>
  );
};
