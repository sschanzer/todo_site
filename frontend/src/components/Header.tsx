import React from "react";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Form from "react-bootstrap/Form";
import { Button, InputGroup } from "react-bootstrap";
import { ITask } from "../App";
import { useEffect, useState } from "react";
import axios from "axios";

// set up to pass pending task and setpending tasks into Header
interface HeaderProps {
  allTasks: ITask[];
  setAllTasks: (allTasks: ITask[]) => void;
}

// export so we can run tests on createTaskResponse
export interface createTaskResponse {
  itemCreated: boolean;
  id: number;
}

export const createTask = async (
  taskTitle: string
): Promise<createTaskResponse> => {
  let response = await axios.post("new_task/", {
    name: taskTitle,
  });
  return response["data"];
};

export const isTaskTitleEmpty = (taskTile: string) => {
  let cleanInput = taskTile.replaceAll(" ", "");
  if (cleanInput.length >= 1 && cleanInput != "") {
    return false;
  } else {
    return true;
  }
};

export const Header: React.FC<HeaderProps> = ({ allTasks, setAllTasks }) => {
  const [newTask, setNewTask] = useState("");
  const [isSubmittedDisabled, setIsSubmittedDisabled] = useState(true);

  useEffect(() => {
    setIsSubmittedDisabled(isTaskTitleEmpty(newTask));
  }, [newTask]);

  const createNewTask = async (
    task: string,
    event?: React.FormEvent<HTMLFormElement>
  ) => {
    event?.preventDefault();
    let response = await createTask(task);
    console.log(response);
    if (response.itemCreated) {
      setAllTasks([
        ...allTasks,
        { id: response.id, title: task, completed: false },
      ]);
      setNewTask("");
    }
  };

  return (
    <Container>
      <Row>
        <Col xs={4}></Col>
        <Col xs={8}>
          <Form
            style={{ position: "relative" }}
            onSubmit={(event) => createNewTask(newTask, event)}
          >
            <Form.Control
              value={newTask}
              id="createTaskInput"
              onChange={(event) => setNewTask(event.target.value)}
            />
            <Button
              variant="success"
              id="createTaskButton"
              disabled={isSubmittedDisabled}
              type="submit"
            >
              +
            </Button>
          </Form>
        </Col>
      </Row>
    </Container>
  );
};

export default Header;
