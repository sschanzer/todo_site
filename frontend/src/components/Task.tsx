import React from "react";
import { ITask } from "../App";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import axios from "axios";
import { Button } from "react-bootstrap";

export interface TaskProps {
  task: ITask;
  selectedTasks: number[];
  setSelectedTasks: (selectedTasks: number[]) => void;
  allTasks: ITask[];
  setAllTasks: (allTasks: ITask[]) => void;
}

export const changeTaskStatus = async (id: number) => {
  let response = await axios.put(`task/${id}/`);
  return response.data.changed;
};

export const deleteTask = async (id: number) => {
  let response = await axios.delete(`task/${id}/`);
  return response.data.success;
};

export const Task: React.FC<TaskProps> = ({
  task,
  allTasks,
  setAllTasks,
  selectedTasks,
  setSelectedTasks,
}) => {
  const changeStatus = async (clicked: boolean, taskToChange: ITask) => {
    let response = await changeTaskStatus(taskToChange["id"]);
    if (response) {
      setAllTasks(allTasks.filter((task) => task != taskToChange));
      taskToChange.completed = clicked;
      setAllTasks([...allTasks]);
    }
  };

  const deleteATask = async (id: number) => {
    let response = await deleteTask(id);
    if (response) {
      setAllTasks(allTasks.filter((task) => task.id !== id));
    }
  };

  return (
    <Row className="task">
      <Col>
        <Form.Check
          id={`taskSelectedBtn${task.id}`}
          type="checkbox"
          onChange={(event) =>
            event.target.checked
              ? setSelectedTasks([...selectedTasks, task.id])
              : setSelectedTasks(selectedTasks.filter((id) => id !== task.id))
          }
        />
      </Col>
      <Col xs={8} id={`task${task.id}`} className="taskTitle">
        {task.title}
      </Col>
      <Col className="checkHolder" xs={2}>
        <Form.Check
          id={`taskCheck${task.id}`}
          type="checkbox"
          className="check"
          checked={task.completed}
          onChange={(event) => changeStatus(event.target.checked, task)}
        />
        <Form.Label for="checkbox">
          {task.completed ? "done" : "pending"}
        </Form.Label>
        <Col xs={1}>
          <Button
            id={`deleteBtn$(task.id)`}
            className="delBtn"
            variant="danger"
            onClick={() => deleteATask(task.id)}
          >
            Delete
          </Button>
        </Col>
      </Col>
    </Row>
  );
};
