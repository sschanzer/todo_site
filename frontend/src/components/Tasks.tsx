import React from "react";
import { myTasks } from "../App";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

interface IdProps {
  task: myTasks;
  selected: number[];
  setSelected: (selected: number[]) => void;
}

export const Tasks: React.FC<IdProps> = ({ task, selected, setSelected }) => {
  return (
    <div className="task">
      <Row>
        <Col xs={8} style={{ display: "flex" }}>
          <p>{task.title}</p>
        </Col>
        <Col xs={4} style={{ display: "flex" }}>
          <Form.Check type="checkbox" label="select" />
        </Col>
      </Row>
    </div>
  );
};
