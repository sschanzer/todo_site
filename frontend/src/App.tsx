import React, { useState, ChangeEvent, useEffect } from "react";
import "bootswatch/dist/darkly/bootstrap.min.css";
import "./App.css";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import { InputGroup } from "react-bootstrap";
import Header from "./components/Header";
import axios from "axios";
import { Task } from "./components/Task";
import { CSRFToken } from "./components/CSRFToken";

export interface ITask {
  id: number;
  title: string;
  completed: boolean;
}

export async function getTasks(): Promise<ITask[]> {
  let response = await axios.get("all_tasks/");
  return response.data.tasks;
}

function App() {
  const [allTasks, setAllTasks] = useState<ITask[]>([]);

  useEffect(() => {
    const getResponse = async () => {
      let response = await getTasks();
      setAllTasks(response);
    };
    getResponse();
  }, []);

  CSRFToken();

  return (
    <div className="App">
      <Header allTasks={allTasks} setAllTasks={setAllTasks} />
      <Row style={{ textAlign: "center" }}>
        <h3>Pending Tasks</h3>
        {allTasks.map((task) =>
          task.completed == false ? <Task task={task} /> : null
        )}
      </Row>

      <Row style={{ textAlign: "center" }}>
        <h3>Completed Tasks</h3>
        {allTasks.map((task) =>
          task.completed == false ? <Task task={task} /> : null
        )}
      </Row>
    </div>
  );
}

export default App;
