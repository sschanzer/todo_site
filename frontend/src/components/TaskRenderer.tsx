import { ITask } from "../App";
import { Task, TaskProps } from "./Task";

interface TaskRendererProps {
  completed: boolean;
  allTasks: ITask[];
  setAllTasks: (pendTasks: ITask[]) => void;
}
export const TaskRenderer: React.FC<TaskRendererProps> = ({
  setAllTasks,
  completed,
  allTasks,
}) => {
  return (
    <>
      {allTasks.map((task) =>
        task.completed === completed ? (
          <Task allTasks={allTasks} setAllTasks={setAllTasks} task={task} />
        ) : null
      )}
    </>
  );
};
