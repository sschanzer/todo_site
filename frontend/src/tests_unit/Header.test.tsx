import { describe, expect, it, vi, Mocked } from "vitest";
import TestRenderer from "react-test-renderer";
import axios from "axios";
import { createTask, deleteTasks } from "../components/Header";
import { Header } from "../components/Header";
import { ITask } from "../App";
import { isTaskTitleEmpty } from "../components/Header";
import { createTaskResponse } from "../components/Header";
import { changeSelectedTasks } from "../components/Header";
import { bool } from "prop-types";

vi.mock("axios");

describe("Header", () => {
  describe("createTask()", () => {
    it("returns a dictionary with itemCreated and id as keys, where itemCreated's value is true and id's value is the task's id", async () => {
      const mockedAxios = axios as Mocked<typeof axios>;
      mockedAxios.post.mockResolvedValue({
        data: { itemCreated: true, id: 1 },
      })<createTaskResponse>;

      const newTasks = await createTask("new task/");

      expect(newTasks).toStrictEqual({ itemCreated: true, id: 1 });
    });
  });

  describe("changeSelectedTasks()", () => {
    it("returns true if multiple tasks are successfully updated", async () => {
      const mockedAxios = axios as Mocked<typeof axios>;
      mockedAxios.put.mockResolvedValue({
        data: { success: true },
      });

      const changedMultipleTasks = await changeSelectedTasks([1, 2, 3]);

      expect(changedMultipleTasks).toBeTruthy();
    });
  });

  describe("isTaskTitleEmpty()", () => {
    it("will return true if input has something other than whitespace", () => {
      const cleanInput = isTaskTitleEmpty("    yes    ");

      expect(cleanInput).toBe(false);
    });

    it("will return false if input has only whitespace", () => {
      const cleanInput = isTaskTitleEmpty("        ");

      expect(cleanInput).toBe(true);
    });
  });

  describe("deleteTasks()", () => {
    it("returns true if tasks are deleted", async () => {
      const mockedAxios = axios as Mocked<typeof axios>;
      mockedAxios.delete.mockResolvedValue({
        data: { success: true },
      });

      const deletedTasks = await deleteTasks([1, 2, 3]);

      expect(deletedTasks).toBe(true);
    });
  });

  it("will create and match snapshot", () => {
    let allTasks: ITask[] = [];
    const setAllTasks = (newAllTasks: ITask[]) => {
      allTasks = newAllTasks;
    };
    let selectedTasks: number[] = [];
    const setSelectedTasks = (taskIdList: number[]) => {
      selectedTasks = taskIdList;
    };
    let showCompleted: boolean = true;
    let showPending: boolean = true;
    const setShowCompleted = (show: boolean) => {
      showCompleted = show;
    };
    const setShowPending = (show: boolean) => {
      showPending = show;
    };

    const myHeader = TestRenderer.create(
      <Header
        allTasks={allTasks}
        setAllTasks={setAllTasks}
        selectedTasks={selectedTasks}
        setSelectedTasks={setSelectedTasks}
        showCompleted={showCompleted}
        showPending={showPending}
        setShowCompleted={setShowCompleted}
        setShowPending={setShowPending}
      />
    );

    expect(myHeader).toMatchSnapshot();
  });
});
