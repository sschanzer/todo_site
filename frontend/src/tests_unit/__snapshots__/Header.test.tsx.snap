// Vitest Snapshot v1

exports[`Header > will create and match snapshot 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-2"
      >
        <button
          className="btn btn-primary btn-sm"
        >
          Delete
        </button>
      </div>
      <div
        className="col-2"
      >
        <button
          className="btn btn-primary btn-sm"
        >
          Status
        </button>
      </div>
      <div
        className="col-1"
      >
        <label
          className="form-check-label"
        >
          P
        </label>
      </div>
      <div
        className="col-2"
      >
        <div
          className="form-check form-switch"
        >
          <input
            className="form-check-input"
            id="flexSwitchCheckDefault"
            role="switch"
            type="checkbox"
          />
        </div>
      </div>
      <div
        className="col-1"
      >
        <label
          className="form-check-label"
        >
          C
        </label>
      </div>
      <div
        className="col-4"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            id="createTaskInput"
            onChange={[Function]}
            placeholder="Task"
            type="text"
            value=""
          />
          <button
            className="btn btn-primary"
            disabled={true}
            id="createTaskButton"
            type="submit"
          >
            +
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
